ring/ring_test.go

```diff
@@ -2234,82 +2342,99 @@ func TestRing_ShuffleShardWithLookback_CorrectnessWithFuzzy(t *testing.T) {
 							nextInstanceID++
-
-							ringDesc.Ingesters[instanceID] = generateRingInstanceWithInfo(instanceID, zoneID, gen.GenerateTokens(128, nil), currTime)
-
-							ring.ringTokens = ringDesc.GetTokens()
-							ring.ringTokensByZone = ringDesc.getTokensByZone()
-							ring.ringInstanceByToken = ringDesc.getTokensInfo()
-							ring.ringZones = getZones(ringDesc.getTokensByZone())
-							if updateOldestRegisteredTimestamp {
-								ring.oldestRegisteredTimestamp = ringDesc.getOldestRegisteredTimestamp()
-							}
-						case r < 90:
-							// Scale down instances by 1. To make tests reproducible we get the instance IDs, sort them
-							// and then get a random index (using the random generator initialized with a constant seed).
-							instanceIDs := make([]string, 0, len(ringDesc.Ingesters))
-							for id := range ringDesc.Ingesters {
-								instanceIDs = append(instanceIDs, id)
+							ringDesc.Ingesters[instanceID] = generateRingInstanceWithInfo(instanceID, zoneID, gen.GenerateTokens(128, ringDesc.GetTokens()), currTime)
+							updateRing()
+							t.Logf("%d: added instance %s", i, instanceID)
+
+						case r < 70:
+							// Scale down instances by 1.
+							idToRemove := getRandomInstanceID(ringDesc.Ingesters, rnd)
+							zone := ringDesc.Ingesters[idToRemove].Zone
+							// Don't remove instance if it is the last instance in the zone,
+							// because sharding works differently for different number of zones.
+							if ring.instancesCountPerZone[zone] <= 1 {
+								t.Logf("%d: not removing last instance %s from zone %s", i, idToRemove, zone)
+								break
 							}
 
-							sort.Strings(instanceIDs)
-
-							idxToRemove := rnd.Intn(len(instanceIDs))
-							idToRemove := instanceIDs[idxToRemove]
 							delete(ringDesc.Ingesters, idToRemove)
-
-							ring.ringTokens = ringDesc.GetTokens()
-							ring.ringTokensByZone = ringDesc.getTokensByZone()
-							ring.ringInstanceByToken = ringDesc.getTokensInfo()
-							ring.ringZones = getZones(ringDesc.getTokensByZone())
-							if updateOldestRegisteredTimestamp {
-								ring.oldestRegisteredTimestamp = ringDesc.getOldestRegisteredTimestamp()
-							}
+							updateRing()
+							t.Logf("%d: removed instance %s", i, idToRemove)
 
 							// Remove the terminated instance from the history.
 							for ringTime, ringState := range history {
 								for idx, desc := range ringState.Instances {
 									// In this simulation instance ID == instance address.
-									if desc.Addr != idToRemove {
-										continue
+									if desc.Addr == idToRemove {
+										ringState.Instances = append(ringState.Instances[:idx], ringState.Instances[idx+1:]...)
+										history[ringTime] = ringState
+										break
 									}
-
-									ringState.Instances = append(ringState.Instances[:idx], ringState.Instances[idx+1:]...)
-									history[ringTime] = ringState
-									break
 								}
 							}
+
+							// Removed instance can't be read-only.
```

Comment:
Uhm... I don't understand this comment. Is it correct?

Reply:
It is. "Removed instance" was already removed from the ring, but we also need to remove it from the `readOnlyInstances` map. Previous version of the test forgot to do that, and then test could try to switch such instance back to "read write" mode, which re-created entry in the ring with no tokens.